waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/AbstractThriftMetastoreClientManager.java

```diff
@@ -0,0 +1,174 @@
+/**
+ * Copyright (C) 2016-2024 Expedia, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.waggledance.client;
+
+import java.io.Closeable;
+import java.net.URI;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.conf.HiveConf.ConfVars;
+import org.apache.hadoop.hive.conf.HiveConfUtil;
+import org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore;
+import org.apache.thrift.TException;
+import org.apache.thrift.transport.TTransport;
+
+import lombok.extern.log4j.Log4j2;
+
+import com.hotels.bdp.waggledance.client.compatibility.HiveCompatibleThriftHiveMetastoreIfaceFactory;
+
+@Log4j2
+public abstract class AbstractThriftMetastoreClientManager implements Closeable {
+
+
+  static final AtomicInteger CONN_COUNT = new AtomicInteger(0);
+  final HiveConf conf;
+  final HiveCompatibleThriftHiveMetastoreIfaceFactory hiveCompatibleThriftHiveMetastoreIfaceFactory;
+  final URI[] metastoreUris;
+  ThriftHiveMetastore.Iface client = null;
+  TTransport transport = null;
+  boolean isConnected = false;
+  // for thrift connects
+  int retries = 5;
+  long retryDelaySeconds = 0;
+
+  final int connectionTimeout;
+  final String msUri;
+
+  AbstractThriftMetastoreClientManager(
+      HiveConf conf,
+      HiveCompatibleThriftHiveMetastoreIfaceFactory hiveCompatibleThriftHiveMetastoreIfaceFactory,
+      int connectionTimeout) {
+    this.conf = conf;
+    this.hiveCompatibleThriftHiveMetastoreIfaceFactory = hiveCompatibleThriftHiveMetastoreIfaceFactory;
+    this.connectionTimeout = connectionTimeout;
+    msUri = conf.getVar(ConfVars.METASTOREURIS);
+
+    if (HiveConfUtil.isEmbeddedMetaStore(msUri)) {
+      throw new RuntimeException("You can't waggle an embedded metastore");
+    }
+
+    // get the number retries
+    retries = HiveConf.getIntVar(conf, ConfVars.METASTORETHRIFTCONNECTIONRETRIES);
+    retryDelaySeconds = conf.getTimeVar(ConfVars.METASTORE_CLIENT_CONNECT_RETRY_DELAY, TimeUnit.SECONDS);
+
+    // user wants file store based configuration
+    if (msUri != null) {
```

Comment:
You should encapsulate this logic in a private method. The constructor should only focus on the main initialisation logic. If the developer is interested in how the mtastoreUris are initialised, it could then read the private method.